statsmodels/regression/censored_model.py

```diff
@@ -0,0 +1,900 @@
+        If False, no censoring in the right tail of the distribution.
+        If a floating point value is given, this is the censoring level on
+        the right.
+
+    Notes
+    -----
+    ystar is the observed variable, that proxies for the unobserved latent
+    variable y.
+
+              y, if a < y < b
+    ystar = { a, if y <= a
+              b, if y >= b
+    """
+    def __init__(self, endog, exog, left=True, right=False):
+        super(Tobit, self).__init__(endog, exog)
+        self._transparams = False
+        # set up censoring
+        self._init_censored(left, right)
+        self._set_loglike()
+
+    def _init_censored(self, left, right):
+        if left is False and right is False:
+            raise ValueError("Must give a censoring level on the right or "
+                             "left.")
+        endog = self.endog
+        exog = self.exog
+        if left is True:
+            left = self.endog.min()
+        elif left is False:
+            left = -np.inf
+        if right is True:
+            right = self.endog.max()
+        elif right is False:
+            right = np.inf
+        self.left, self.right = left, right
+
+        # do this all here, so that we only have to do the indexing once
+        left_idx = endog <= left # do we need to attach these?
+        right_idx = endog >= right
+        self._left_endog = endog[left_idx]
+        self.n_lcens = left_idx.sum()
+        self.n_rcens = right_idx.sum()
+        self._right_endog = endog[right_idx]
+        self._left_exog = exog[left_idx]
+        self._right_exog = exog[right_idx]
+        center_idx = ~np.logical_or(left_idx, right_idx)
+        self._center_endog = endog[center_idx]
+        self._center_exog = exog[center_idx]
+
+    def _set_loglike(self):
+        """
+        Use this so we don't have a lot of if checks in each loglike call
+        """
+        left, right = self.left, self.right
+        if np.isfinite(left) and np.isinf(right): # left-censored
+            self._loglike = self._loglike_left
+            self._score = self._score_left
+            #self.hessian_ = self._hessian_left
+        elif np.isinf(left) and np.isfinite(right): # right-censored
+            self._loglike = self._loglike_right
+            self._score = self._score_right
+            #self.hessian = self._hessian_right
+        else: # left and right censored
+            self._loglike = self._loglike_both
+            self._score = self._score_both
+            #self.hessian = self._hessian_both
+
+    @transform2(_exponentiate_sigma) # keeps std. dev. positive
+    def loglike(self, params):
+        """
+        Notes
+        -----
+        Assumes sigma is the last parameter
+        """
+        params, sigma = params[:-1], params[-1]
+        return self._loglike(params, sigma)
+
+    @transform2(_exponentiate_sigma)
+    def score(self, params):
+        params, sigma = params[:-1], params[-1]
+        return self._score(params, sigma)
```

Comment:
too many helper functions and methods, makes it difficult to follow code.
params is not uniquely defined, params including scale/sigma, and params without scale/sigma